flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -113,23 +113,36 @@ class AbstractArrayBox {
     return lbounds;
   }
 
+  /// If the entity is described by a box (e.g. it is a dummy, allocatable, or
+  /// pointer) The information in the source box that is also present in other
+  /// fields (e.g the extents) is the same, except the lower bounds of the
+  /// sourceBox that should be ignored.
+  mlir::Value getSourceBox() const { return sourceBox; }
+
   bool lboundsAllOne() const { return lbounds.empty(); }
   std::size_t rank() const { return extents.size(); }
+  /// Is this entity guaranteed to be contiguous at compile time ?
+  /// If not, the memory layout is described by sourceBox.
+  bool isContiguous() const;
 
 protected:
   llvm::SmallVector<mlir::Value, 4> extents;
   llvm::SmallVector<mlir::Value, 4> lbounds;
+  mlir::Value sourceBox;
```

Comment:
Is this the right place for this? I wonder if it would be helpful to have boxed-array vs. unboxed-array explicit in this (C++) type hierarchy.

Reply:
I am a bit reluctant in duplicating the C++ hierarchy just for that, but I think this should at least be the main address (addr field from AbstractBox) for non-contiguous entities, because we most likely do not want people to directly use the base address by mistake when using fir::getBase(). The reason I did not is that I think most current code calling fir::getBase() expects addresses, and I need to audit this. But If that sounds reasonable to you, I can do that now.

Note that for non-contiguous entities, there might still be reasons to keep track of the descriptor it came from to allow descriptor propagation and such (There would probably be other ways find the fir.box back/keep track of it, but placing it in the SymbolBox seems the safer to me, though in those case, we probably do not want that to be the "addr" field).